File: portal/tools/pocwu.nim
# ─── pocwu.nim — Nim utility for project maintenance ─────────────────────
#
# Provides project health checks, file counting, and orchestration.
#
# Build & Run:
#   nim compile --run tools/pocwu.nim
#
# ──────────────────────────────────────────────────────────────────────────

import std/[os, strformat, strutils, times]

# ─── Types ──────────────────────────────────────────────────────────────

type
  LangInfo = object
    name: string
    exts: seq[string]
    count: int

  FileCheck = object
    path: string
    label: string

# ─── Helpers ────────────────────────────────────────────────────────────

proc countFiles(root: string, exts: seq[string]): int =
  for (path, dirs, files) in walkDir(root):
    if dirs.len > 0:
      for d in dirs:
        let full = path / d
        if d notin ["node_modules", "dist", "target", ".git"]:
          result += countFiles(full, exts)
    for f in files:
      let (_, name, ext) = splitFile(f)
      if ext in exts or name in exts:
        result += 1

proc formatCount(n: int): string =
  ($n).align(4)

# ─── Main ──────────────────────────────────────────────────────────────

proc main() =
  let root = getCurrentDir()
  let now = now()
  echo ""
  echo "═══════════════════════════════════════════════"
  echo "  OpenCodeABs/UX — Nim Project Utility"
  echo &"  {now.format(\"yyyy-MM-dd HH:mm:ss\")}"
  echo "═══════════════════════════════════════════════"
  echo ""

  # ─── Language counts ────────────────────────────────

  var langs = @[
    LangInfo(name: "TypeScript",   exts: @[".ts", ".tsx"]),
    LangInfo(name: "JavaScript",   exts: @[".js"]),
    LangInfo(name: "CSS",          exts: @[".css"]),
    LangInfo(name: "JSON",         exts: @[".json"]),
    LangInfo(name: "HTML",         exts: @[".html"]),
    LangInfo(name: "Rust",         exts: @[".rs"]),
    LangInfo(name: "Python",       exts: @[".py"]),
    LangInfo(name: "Go",           exts: @[".go"]),
    LangInfo(name: "C",            exts: @[".c", ".h"]),
    LangInfo(name: "GLSL",         exts: @[".vert", ".frag"]),
    LangInfo(name: "Lua",          exts: @[".lua"]),
    LangInfo(name: "Ruby",         exts: @[".rb"]),
    LangInfo(name: "Zig",          exts: @[".zig"]),
    LangInfo(name: "Kotlin",       exts: @[".kt", ".kts"]),
    LangInfo(name: "Nim",          exts: @[".nim"]),
    LangInfo(name: "Shell",        exts: @[".sh"]),
    LangInfo(name: "SQL",          exts: @[".sql"]),
    LangInfo(name: "TOML",         exts: @[".toml"]),
    LangInfo(name: "PowerShell",   exts: @[".ps1"]),
    LangInfo(name: "YAML",         exts: @[".yml", ".yaml"]),
    LangInfo(name: "Dockerfile",   exts: @["Dockerfile"]),
    LangInfo(name: "Makefile",     exts: @["Makefile", ".mk"]),
  ]

  var total = 0
  let maxNameLen = 15

  echo &"  {'Language'.align(maxNameLen)}  Files"
  echo &"  {'────────'.align(maxNameLen)}  ─────"
  for lang in langs.mitems:
    lang.count = countFiles(root, lang.exts)
    total += lang.count
    echo &"  {lang.name.align(maxNameLen)}  {lang.count.formatCount}"
  echo &"  {'────────'.align(maxNameLen)}  ─────"
  echo &"  {'TOTAL'.align(maxNameLen)}  {total.formatCount}"
  echo ""

  # ─── File checks ──────────────────────────────────

  let checks = @[
    FileCheck(path: "Makefile", label: "Build automation"),
    FileCheck(path: "Dockerfile", label: "Container build"),
    FileCheck(path: "package.json", label: "Node.js project"),
    FileCheck(path: "tsconfig.json", label: "TypeScript config"),
    FileCheck(path: "wrangler.toml", label: "Wrangler config"),
    FileCheck(path: "deploy.ps1", label: "PowerShell deploy"),
    FileCheck(path: "scripts/deploy.sh", label: "Bash deploy"),
    FileCheck(path: "scripts/generate_cities.py", label: "Python generator"),
    FileCheck(path: "scripts/pocwu.lua", label: "Lua config"),
    FileCheck(path: "scripts/preview.rb", label: "Ruby preview"),
    FileCheck(path: "native/globe_math.h", label: "C math header"),
    FileCheck(path: "native/globe_math.c", label: "C math impl"),
    FileCheck(path: "native/globe.zig", label: "Zig math module"),
    FileCheck(path: "rswasm-globe-physics/Cargo.toml", label: "Rust project"),
    FileCheck(path: "cmd/pocwu/main.go", label: "Go CLI"),
    FileCheck(path: "shaders/globe.vert", label: "GLSL vertex shader"),
    FileCheck(path: "shaders/globe.frag", label: "GLSL fragment shader"),
    FileCheck(path: "tools/pocwu.main.kts", label: "Kotlin script"),
    FileCheck(path: "tools/pocwu.nim", label: "Nim utility"),
    FileCheck(path: "schema.sql", label: "SQL schema"),
  ]

  echo "📋 File existence checks:"
  echo ""
  var allGood = true
  for check in checks:
    if fileExists(check.path):
      echo &"  ✅ {check.label} ({check.path})"
    else:
      echo &"  ❌ {check.label} ({check.path} — MISSING)"
      allGood = false

  echo ""
  if allGood:
    echo "✅ All files present — polyglot architecture healthy!"
  else:
    echo "⚠️  Some files missing — run 'make polyglot' for details"

  echo ""

when isMainModule:
  main()

File: portal/scripts/pocwu.lua
-- ─── pocwu.lua — Project build configuration script ──────────────────────
--
-- Provides build automation functions that can be used standalone or
-- integrated with the Makefile / shell pipeline.
--
-- Usage:
--   lua scripts/pocwu.lua build
--   lua scripts/pocwu.lua check
--   lua scripts/pocwu.lua count
--
-- ──────────────────────────────────────────────────────────────────────────

local M = {}

-- ─── Configuration ─────────────────────────────────────────────────────────

M.config = {
    project_name = "OpenCodeABs/UX",
    version = "0.1.0",
    src_dir = "src",
    dist_dir = "dist",
    langs = {
        "TypeScript", "JavaScript", "CSS", "JSON", "HTML",
        "Rust", "Python", "Go", "C", "GLSL", "Lua", "Ruby",
        "Zig", "Kotlin", "Nim", "Shell", "SQL", "TOML",
        "PowerShell", "YAML", "Dockerfile", "Makefile"
    }
}

-- ─── File Helpers ─────────────────────────────────────────────────────────

local function file_exists(path)
    local f = io.open(path, "r")
    if f then
        f:close()
        return true
    end
    return false
end

local function read_file(path)
    local f = io.open(path, "r")
    if not f then return nil end
    local content = f:read("*all")
    f:close()
    return content
end

-- ─── Build Check ──────────────────────────────────────────────────────────

function M.check()
    print("🔍 " .. M.config.project_name .. " — Build Check")
    print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")

    local checks = {
        {"wrangler.toml", "Wrangler config"},
        {"package.json", "Node.js project"},
        {"tsconfig.json", "TypeScript config"},
        {"Makefile", "Build automation"},
        {"Dockerfile", "Container build"},
        {"deploy.ps1", "PowerShell deploy"},
        {"scripts/deploy.sh", "Bash deploy"},
        {"scripts/generate_cities.py", "Python generator"},
        {"native/globe_math.h", "C math header"},
        {"native/globe_math.c", "C math impl"},
        {"rswasm-globe-physics/Cargo.toml", "Rust project"},
        {"rswasm-globe-physics/src/lib.rs", "Rust impl"},
        {"cmd/pocwu/main.go", "Go CLI"},
        {"shaders/globe.vert", "GLSL vertex"},
        {"shaders/globe.frag", "GLSL fragment"},
    }

    local passed, failed = 0, 0
    for _, check in ipairs(checks) do
        local path, label = check[1], check[2]
        if file_exists(path) then
            print("  ✅ " .. label .. " (" .. path .. ")")
            passed = passed + 1
        else
            print("  ❌ " .. label .. " (" .. path .. " — MISSING)")
            failed = failed + 1
        end
    end

    print("")
    print("Result: " .. passed .. " passed, " .. failed .. " failed")
    return failed == 0
end

-- ─── Language File Count ──────────────────────────────────────────────────

function M.count()
    print("📊 Language File Count")
    print("━━━━━━━━━━━━━━━━━━━━━")

    local patterns = {
        TypeScript = {".ts", ".tsx"},
        JavaScript = {".js"},
        CSS = {".css"},
        JSON = {".json"},
        HTML = {".html"},
        Rust = {".rs"},
        Python = {".py"},
        Go = {".go"},
        C = {".c", ".h"},
        GLSL = {".vert", ".frag"},
        Lua = {".lua"},
        Ruby = {".rb"},
        Zig = {".zig"},
        Kotlin = {".kt", ".kts"},
        Nim = {".nim"},
        Shell = {".sh"},
        SQL = {".sql"},
        TOML = {".toml"},
        PowerShell = {".ps1"},
        YAML = {".yml", ".yaml"},
        Dockerfile = {"Dockerfile"},
        Makefile = {"Makefile", ".mk"},
    }

    -- Use a simple recursive directory walk
    local function count_files(dir, exts)
        local count = 0
        local handle = io.popen('dir /s /b "' .. dir .. '" 2>nul')
        if not handle then return 0 end
        local result = handle:read("*all")
        handle:close()

        for line in result:gmatch("[^\r\n]+") do
            for _, ext in ipairs(exts) do
                if line:sub(-#ext) == ext then
                    count = count + 1
                    break
                end
            end
        end
        return count
    end

    local total = 0
    for name, exts in pairs(patterns) do
        local c = count_files(".", exts)
        if c > 0 then
            print(string.format("  %-15s %d files", name, c))
            total = total + c
        end
    end
    print("  ─────────────────────")
    print(string.format("  %-15s %d files", "TOTAL", total))
    return total
end

-- ─── Build (delegate to npm) ──────────────────────────────────────────────

function M.build()
    print("🔨 Building frontend...")
    local result = os.execute("npm run build")
    if result then
        print("✅ Build complete")
    else
        print("❌ Build failed")
    end
    return result
end

-- ─── Main ─────────────────────────────────────────────────────────────────

local commands = {
    check = M.check,
    count = M.count,
    build = M.build,
}

local cmd = arg and arg[1]
if cmd and commands[cmd] then
    local ok = commands[cmd]()
    if ok == false then
        os.exit(1)
    end
else
    print("pocwu.lua — OpenCodeABs/UX build config")
    print("")
    print("Usage: lua scripts/pocwu.lua <command>")
    print("")
    print("Commands:")
    print("  check    Verify all project files exist")
    print("  count    Count files by language")
    print("  build    Build the frontend")
end

return M
